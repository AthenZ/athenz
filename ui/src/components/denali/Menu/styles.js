/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { css } from 'emotion';
import { cssFontStyles } from '../styles/fonts';
import { colors } from '../styles/colors';
import { cssDropShadow } from '../styles/drop-shadow';

export const makeCssPopperBox = (props) => css`
    ${cssDropShadow};
    ${cssFontStyles.default};
    background-color: ${colors.white};
    border: 2px solid ${colors.border};
    border-radius: 4px;
    color: ${colors.black};
    cursor: default;
    line-height: 1.4;
    margin: 0;
    max-width: ${props.basic ? '400px' : undefined};
    min-width: 20px;
    padding: ${props.basic ? '6px 10px' : 0};
    transition: ${!props.noanim && 'opacity 0.25s ease-in'};
    z-index: 999;
    label: denali-menu;
    /* Shift the popper to allow space for arrow */
    &[data-placement*='bottom'] {
        margin-top: 8px;
    }
    &[data-placement*='top'] {
        margin-bottom: 8px;
    }
    &[data-placement*='left'] {
        margin-right: 8px;
    }
    &[data-placement*='right'] {
        margin-left: 8px;
    }

    /**
   * Invisible padding between trigger and main content (around the arrow) so
   * user can slide from one to the other without the menu closing.
   */
    &::after {
        background: transparent;
        content: '';
        position: absolute;
    }
    &[data-placement*='top']::after {
        bottom: -12px;
        height: 12px;
        width: 100%;
    }
    &[data-placement*='bottom']::after {
        height: 12px;
        top: -12px;
        width: 100%;
    }
    &[data-placement*='left']::after {
        right: -12px;
        height: 100%;
        width: 12px;
    }
    &[data-placement*='right']::after {
        left: -12px;
        height: 100%;
        width: 12px;
    }
`;

export const cssArrow = css`
    height: 1rem;
    position: absolute;
    width: 1rem;
    label: menu-arrow;

    &::before,
    &::after {
        border-style: solid;
        content: '';
        display: block;
        height: 0;
        margin: auto;
        position: absolute;
        width: 0;
    }

    &[data-placement*='bottom'] {
        left: 0;
        margin-top: -6px;
        top: -8px;
        width: 1rem;
    }

    &[data-placement*='bottom']::before {
        border-color: transparent transparent ${colors.grey400} transparent;
        border-width: 0 8px 8px 8px;
        position: absolute;
        top: 4px;
    }

    &[data-placement*='bottom']::after {
        border-color: transparent transparent ${colors.white} transparent;
        border-width: 0 8px 8px 8px;
        top: 7px;
    }

    &[data-placement*='top'] {
        bottom: 0;
        height: 1rem;
        left: 0;
        margin-bottom: -1rem;
        width: 1rem;
    }

    &[data-placement*='top']::before {
        border-color: ${colors.grey400} transparent transparent transparent;
        border-width: 8px 8px 0 8px;
        position: absolute;
        top: 2px;
    }

    &[data-placement*='top']::after {
        border-color: ${colors.white} transparent transparent transparent;
        border-width: 8px 8px 0 8px;
        top: -1px;
    }

    &[data-placement*='right'] {
        height: 1rem;
        left: 0;
        margin-left: -1rem;
        width: 1rem;
    }

    &[data-placement*='right']::before {
        border-color: transparent ${colors.grey400} transparent transparent;
        border-width: 8px 8px 8px 0;
        left: 5px;
    }

    &[data-placement*='right']::after {
        border-color: transparent ${colors.white} transparent transparent;
        border-width: 8px 8px 8px 0;
        left: 8px;
        top: 0;
    }

    &[data-placement*='left'] {
        height: 1rem;
        margin-right: -1rem;
        right: 0;
        width: 1rem;
    }

    &[data-placement*='left']::before {
        border-color: transparent transparent transparent ${colors.grey400};
        border-width: 8px 0 8px 8px;
        left: 2px;
    }

    &[data-placement*='left']::after {
        border-color: transparent transparent transparent ${colors.white};
        border-width: 8px 0 8px 8px;
        left: -1px;
        top: 0;
    }
`;
