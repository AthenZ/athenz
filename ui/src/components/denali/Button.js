/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { cx, css } from 'emotion';
import { rgba } from 'polished';
import { colors } from './styles/colors';
import { cssFontFamilies, cssFontWeights } from './styles/fonts';

// NOTE: Psuedo class priority:
// Link Visited Hover Focus Active
// "Lord Vader Hates Furry Animals"

const buttonBase = (props) => css`
    border: none;
    border-radius: 2px;
    box-shadow: 0 0 0 1px transparent inset, 0 0 0 0 ${colors.grey500} inset;
    box-sizing: border-box;
    cursor: pointer;
    display: inline-block;
    ${cssFontFamilies.default};
    ${cssFontWeights.normal};
    font-size: ${props.size === 'large'
        ? '18px'
        : props.size === 'regular'
        ? '14px'
        : '12px'};
    line-height: 1;
    margin: 5px;
    min-width: ${props.size === 'small' ? '70px' : undefined};
    outline: 0;
    padding: ${props.size === 'large'
        ? '12px 28px'
        : props.size === 'regular'
        ? '10px 24px'
        : '8px 18px'};
    text-align: center;
    text-shadow: none;
    text-transform: none;
    transition: ${!props.noanim ? 'all 0.2s ease-in' : undefined};
    vertical-align: baseline;
    white-space: nowrap;
    label: button;
    &:first-of-type {
        margin-left: 0;
    }
    &:disabled {
        background: ${props.dark
            ? rgba(colors.white, 0.1)
            : rgba(colors.grey800, 0.1)};
        color: ${props.dark
            ? rgba(colors.white, 0.2)
            : rgba(colors.grey800, 0.2)};
        cursor: not-allowed;
    }
`;

const buttonVariants = {
    primary: (props) => css`
        background: linear-gradient(
            to right,
            ${props.active ? colors.brand700 : colors.brand600},
            ${props.active ? colors.brand800 : colors.brand700}
        );
        color: ${colors.white};
        &:hover:not(:disabled) {
            background: linear-gradient(
                to right,
                ${colors.brand700},
                ${colors.brand800}
            );
        }
    `,
    secondary: (props) => css`
        background: ${props.dark
            ? props.active
                ? `linear-gradient(to right, ${rgba(
                      colors.brand700,
                      0.3
                  )}, ${rgba(colors.brand800, 0.3)})`
                : 'transparent'
            : props.active
            ? colors.brand200
            : 'transparent'};
        box-shadow: 0 0 0 1px ${colors.brand600} inset;
        color: ${props.dark ? colors.white : colors.brand600};
        &:disabled {
            box-shadow: ${props.dark
                ? `0 0 0 1px ${rgba(colors.white, 0.2)} inset`
                : `0 0 0 1px ${rgba(colors.grey800, 0.1)} inset`};
        }
        &:hover:not(:disabled) {
            background: ${props.dark
                ? `linear-gradient(to right, ${rgba(
                      colors.brand700,
                      0.3
                  )}, ${rgba(colors.brand800, 0.3)})`
                : colors.brand200};
        }
    `,
    danger: () => css`
        background: ${colors.red500};
        color: ${colors.white};
        &:hover:not(:disabled) {
            background: ${colors.red600};
        }
    `,
};

class Button extends React.PureComponent {
    render() {
        const {
            active,
            className,
            danger,
            dark,
            noanim,
            secondary,
            size,
            ...rest
        } = this.props;

        const cssProps = { active, dark, noanim, size };

        const makeCssButton = (props) => css`
            ${buttonBase(props)};
            ${(danger
                ? buttonVariants.danger
                : secondary
                ? buttonVariants.secondary
                : buttonVariants.primary)(props)};
        `;
        const cssButton = makeCssButton(cssProps);

        const classes = cx(cssButton, 'denali-button', className);

        return <button className={classes} {...rest} />;
    }
}

Button.propTypes = {
    /** Set active state */
    active: PropTypes.bool,
    /** Additonal class to apply to the button */
    className: PropTypes.string,
    /** Error color scheme */
    danger: PropTypes.bool,
    /** Dark color scheme (can be used in conjunction with 'secondary') */
    dark: PropTypes.bool,
    /** Disable the button (no clicks) */
    disabled: PropTypes.bool,
    /** Forward React ref to the `<input>` element */
    innerRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    /** Name of the `button` element */
    name: PropTypes.string,
    /** Disable animations / transitions */
    noanim: PropTypes.bool,
    /** Secondary color scheme */
    secondary: PropTypes.bool,
    /** Size of buttons (default is regular) */
    size: PropTypes.oneOf(['large', 'regular', 'small']),
    /** <button> type attribute */
    type: PropTypes.oneOf(['button', 'submit', 'reset']),
    /** <button> value attribute */
    value: PropTypes.any,
    /** Gets called when the button is clicked */
    onClick: PropTypes.func,
};

Button.defaultProps = {
    active: false,
    danger: false,
    dark: false,
    disabled: false,
    noanim: false,
    size: 'regular',
    secondary: false,
};

export default Button;
