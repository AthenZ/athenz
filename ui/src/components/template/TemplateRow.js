/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import styled from '@emotion/styled';
import Icon from '../denali/icons/Icon';
import { colors } from '../denali/styles';
import DateUtils from '../utils/DateUtils';
import Menu from '../denali/Menu/Menu';
import TemplateDescription from './TemplateDescription';
import Button from '../denali/Button';
import ApplyTemplate from './ApplyTemplate';
import RequestUtils from '../utils/RequestUtils';

const TdStyled = styled.td`
    background-color: ${(props) => props.color};
    text-align: ${(props) => props.align};
    padding: 5px 0 5px 15px;
    vertical-align: middle;
    word-break: break-all;
`;

export default class TemplateRow extends React.Component {
    constructor(props) {
        super(props);
        this.toggleDescription = this.toggleDescription.bind(this);
        this.toggleApplyTemplateNoKeyword =
            this.toggleApplyTemplateNoKeyword.bind(this);
        this.toggleApplyTemplate = this.toggleApplyTemplate.bind(this);
        this.onCancelUpdateTemplate = this.onCancelUpdateTemplate.bind(this);
        this.reloadTemplates = this.reloadTemplates.bind(this);
        this.state = {
            templateDesc: '',
            applyTemplate: false,
            showSuccess: false,
        };
    }

    toggleDescription() {
        if (this.state.description) {
            this.setState({ description: null });
        } else {
            this.setState({
                description: this.props.templateDesc,
            });
        }
    }

    toggleApplyTemplate() {
        this.setState({
            showUpdate: true,
            errorMessage: null,
        });
    }

    toggleApplyTemplateNoKeyword() {
        let params = {
            name: this.props.domain,
            domainTemplate: { templateNames: [this.props.templateName] },
        };
        this.props.api
            .updateTemplate(params, this.props._csrf)
            .then(() => {
                this.reloadTemplates();
            })
            .catch((err) => {
                this.setState({
                    errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                });
                this.props.showError(RequestUtils.xhrErrorCheckHelper(err));
            });
    }

    reloadTemplates() {
        this.props.onsubmit(
            `Successfully updated template ${this.props.templateName}`
        );
    }

    closeModal() {
        this.setState({ successService: null });
    }

    onCancelUpdateTemplate() {
        this.setState({
            showUpdate: false,
            deleteServiceName: null,
        });
    }

    render() {
        const left = 'left';
        const center = 'center';
        const color = this.props.color;
        const currentVersion = this.props.currentVersion;
        const latestVersion = this.props.latestVersion;
        const keywordsToReplace = this.props.keywordsToReplace;
        let buttonText = 'Update';
        if (currentVersion == 'N/A') {
            buttonText = 'Onboard';
        }
        let row = [];
        const templateName = this.props.templateName;

        let applyTemplate = this.state.showUpdate ? (
            <ApplyTemplate
                api={this.props.api}
                domain={this.props.domain}
                onSubmit={this.onClickUpdateTemplate}
                onCancel={this.onCancelUpdateTemplate}
                _csrf={this.props._csrf}
                showApplyTemplate={this.state.showUpdate}
                keywords={keywordsToReplace}
                templateName={templateName}
                reloadTemplatePage={this.reloadTemplates}
                title={
                    this.props.currentVersion
                        ? 'Update Template'
                        : 'Onboard Template'
                }
            />
        ) : (
            ''
        );

        row.push(
            <tr key={templateName} data-testid='template-row'>
                <TdStyled color={color} align={left}>
                    {templateName}
                </TdStyled>
                <TdStyled color={color} align={center}>
                    <span>
                        <Icon
                            icon={'information-circle'}
                            onClick={this.toggleDescription}
                            color={colors.icons}
                            isLink
                            size={'1.25em'}
                            verticalAlign={'text-bottom'}
                        />
                    </span>
                </TdStyled>
                <TdStyled color={color} align={center}>
                    {currentVersion}
                </TdStyled>
                <TdStyled color={color} align={center}>
                    {latestVersion}
                </TdStyled>
                <TdStyled color={color} align={center}>
                    {this.props.timestamp}
                </TdStyled>
                <TdStyled color={color} align={center}>
                    <Button
                        secondary={
                            currentVersion == 'N/A' ||
                            currentVersion == latestVersion
                                ? true
                                : false
                        }
                        onClick={
                            !keywordsToReplace
                                ? this.toggleApplyTemplateNoKeyword
                                : this.toggleApplyTemplate
                        }
                        size={'small'}
                        type={'submit'}
                    >
                        {buttonText}
                    </Button>
                    {applyTemplate}
                </TdStyled>
            </tr>
        );

        if (this.state.description) {
            row.push(
                <tr key={templateName + this.props.templateDesc}>
                    <TemplateDescription
                        color={this.props.color}
                        description={this.state.description}
                        api={this.api}
                    />
                </tr>
            );
        }
        return row;
    }
}
