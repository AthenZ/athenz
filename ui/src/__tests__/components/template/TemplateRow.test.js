/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import {fireEvent, render, waitForElement} from '@testing-library/react';
import TemplateRow from '../../../components/template/TemplateRow';
import { colors } from '../../../components/denali/styles';

describe('TemplateRow', () => {
    function onClickUpdateTemplate() {
        return false;
    }

    function returnTemplateName(){
        return "aws";
    }

    it('should render templateRow', () => {
        const color = colors.row;
        const currentVersion = 0;
        const latestVersion = 1;
        const keywordsToReplace = "";
        const { getByTestId } = render(
            <table>
                <tbody>
                <TemplateRow currentVersion={currentVersion}
                             latestVersion={latestVersion}
                             keywordsToReplace={keywordsToReplace}
                             color={color} />
                </tbody>
            </table>
        );
        const templateRow = getByTestId('template-row');
        expect(templateRow).toMatchSnapshot();
    });

    it('should render without pop up', () => {
        const color = colors.row;
        const currentVersion = 2;
        const latestVersion = 2;
        const keywordsToReplace = "";
        const templateDescription = "testing template description";
        const { getByTestId } = render(
            <table>
                <tbody>
                <TemplateRow currentVersion={currentVersion}
                             latestVersion={latestVersion}
                             keywordsToReplace={keywordsToReplace}
                             color={color}
                             description={templateDescription}/>
                </tbody>
            </table>
        );
        const templateRow = getByTestId('template-row');
        expect(templateRow).toMatchSnapshot();
    });

    it('should render with description', async () => {
        const color = colors.row;
        const currentVersion = 2;
        const latestVersion = 2;
        const keywordsToReplace = "";
        const templateDescription = "testing template description";
        const { getByTitle,  getByTestId,  } = render(
            <table>
                <tbody>
                <TemplateRow currentVersion={currentVersion}
                             latestVersion={latestVersion}
                             keywordsToReplace={keywordsToReplace}
                             color={color}
                             description={templateDescription}/>
                </tbody>
            </table>
        );
        await waitForElement(() =>
            fireEvent.click(getByTitle('information-circle'))
        );

        expect(
            await waitForElement(() => getByTestId('template-row'))
        ).toMatchSnapshot();
    });


    it('should render templaterow with pop up', async () => {
        const color = colors.row;
        const currentVersion = 2;
        const latestVersion = 2;
        const keywordsToReplace = "";
        const templateDescription = "testing template description";
        const templateName= 'aws';
        const domain= 'testdom';

        let params = {
            name: domain,
            domainTemplate: { templateNames: [templateName] },
        };

        let toReturn =
            [{
                "templateName": "aws",
                "description": "AWS access template",
                "currentVersion": 4,
                "latestVersion": 1,
                "timestamp": "2020-04-28T00:00:00.000Z",
                "autoUpdate": false
            }];

        const api = {
            updateTemplate: function(params, csrf) {
                return new Promise((resolve, reject) => {
                    resolve(toReturn); reject("error");
                });
            },
        };

        const {getByTestId } = render(
            <table>
                <tbody>
                <TemplateRow currentVersion={currentVersion}
                             latestVersion={latestVersion}
                             keywordsToReplace={keywordsToReplace}
                             color={color}
                             description={templateDescription}
                             api={api}
                             errorMessage={null}
                             templateName={templateName}
                             templateDesc={templateDescription}
                             onClickUpdateTemplate={onClickUpdateTemplate()}
                             domain={domain}
                             onsubmit={returnTemplateName()}
                             onCancel={onClickUpdateTemplate()}

                />
                </tbody>
            </table>
        );
        const templateRow = getByTestId('template-row');
        expect(
            await waitForElement(() => getByTestId('template-row'))
        ).toMatchSnapshot();
    });
});
