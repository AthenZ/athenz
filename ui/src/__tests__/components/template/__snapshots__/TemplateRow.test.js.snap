// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateRow should render templateRow 1`] = `
.emotion-1 {
  fill: #188fff;
  cursor: pointer;
  vertical-align: text-bottom;
}

.emotion-6 {
  border: none;
  border-radius: 2px;
  box-shadow: 0 0 0 1px transparent inset,0 0 0 0 #d5d5d5 inset;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  font-family: Helvetica,Arial,sans-serif;
  font-weight: 300;
  font-size: 12px;
  line-height: 1;
  margin: 5px;
  min-width: 70px;
  outline: 0;
  padding: 8px 18px;
  text-align: center;
  text-shadow: none;
  text-transform: none;
  -webkit-transition: all 0.2s ease-in;
  transition: all 0.2s ease-in;
  vertical-align: baseline;
  white-space: nowrap;
  background: linear-gradient( to right, #3697f2, #3570f4 );
  color: #fff;
}

.emotion-6:first-of-type {
  margin-left: 0;
}

.emotion-6:disabled {
  background: rgba(48,48,48,0.1);
  color: rgba(48,48,48,0.2);
  cursor: not-allowed;
}

.emotion-6:hover:not(:disabled) {
  background: linear-gradient( to right, #3570f4, #3448f7 );
}

.emotion-0 {
  background-color: #3570f40D;
  text-align: left;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
}

.emotion-2 {
  background-color: #3570f40D;
  text-align: center;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
}

<tr
  data-testid="template-row"
>
  <td
    class="emotion-0"
    color="#3570f40D"
  />
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    <span>
      <svg
        class="emotion-1"
        data-testid="icon"
        height="1.25em"
        viewBox="0 0 1024 1024"
        width="1.25em"
      >
        <title>
          information-circle
        </title>
        <path
          d="M512 42.667c-259.206 0-469.333 210.128-469.333 469.333s210.128 469.333 469.333 469.333c259.206 0 469.333-210.128 469.333-469.333v0c0.003-0.635 0.005-1.387 0.005-2.138 0-258.027-209.173-467.2-467.2-467.2-0.752 0-1.504 0.002-2.255 0.005l0.117-0zM512 896c-212.077 0-384-171.923-384-384s171.923-384 384-384c212.077 0 384 171.923 384 384v0c-1.203 211.592-172.408 382.797-383.885 383.999l-0.115 0.001z"
        />
        <path
          d="M512 426.667c-23.564 0-42.667 19.103-42.667 42.667v0 256c0 23.564 19.103 42.667 42.667 42.667s42.667-19.103 42.667-42.667v0-256c0-23.564-19.103-42.667-42.667-42.667v0z"
        />
        <path
          d="M554.667 298.667c0 23.564-19.103 42.667-42.667 42.667s-42.667-19.103-42.667-42.667c0-23.564 19.103-42.667 42.667-42.667s42.667 19.103 42.667 42.667z"
        />
      </svg>
    </span>
  </td>
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    0
  </td>
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    1
  </td>
  <td
    class="emotion-2"
    color="#3570f40D"
  />
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    <button
      class="emotion-6 denali-button"
      type="submit"
    >
      Update
    </button>
    
  </td>
</tr>
`;

exports[`TemplateRow should render templaterow with pop up 1`] = `
.emotion-1 {
  fill: #188fff;
  cursor: pointer;
  vertical-align: text-bottom;
}

.emotion-6 {
  border: none;
  border-radius: 2px;
  box-shadow: 0 0 0 1px transparent inset,0 0 0 0 #d5d5d5 inset;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  font-family: Helvetica,Arial,sans-serif;
  font-weight: 300;
  font-size: 12px;
  line-height: 1;
  margin: 5px;
  min-width: 70px;
  outline: 0;
  padding: 8px 18px;
  text-align: center;
  text-shadow: none;
  text-transform: none;
  -webkit-transition: all 0.2s ease-in;
  transition: all 0.2s ease-in;
  vertical-align: baseline;
  white-space: nowrap;
  background: transparent;
  box-shadow: 0 0 0 1px #3697f2 inset;
  color: #3697f2;
}

.emotion-6:first-of-type {
  margin-left: 0;
}

.emotion-6:disabled {
  background: rgba(48,48,48,0.1);
  color: rgba(48,48,48,0.2);
  cursor: not-allowed;
}

.emotion-6:disabled {
  box-shadow: 0 0 0 1px rgba(48,48,48,0.1) inset;
}

.emotion-6:hover:not(:disabled) {
  background: #d7e2fd;
}

.emotion-0 {
  background-color: #3570f40D;
  text-align: left;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
}

.emotion-2 {
  background-color: #3570f40D;
  text-align: center;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
}

<tr
  data-testid="template-row"
>
  <td
    class="emotion-0"
    color="#3570f40D"
  >
    aws
  </td>
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    <span>
      <svg
        class="emotion-1"
        data-testid="icon"
        height="1.25em"
        viewBox="0 0 1024 1024"
        width="1.25em"
      >
        <title>
          information-circle
        </title>
        <path
          d="M512 42.667c-259.206 0-469.333 210.128-469.333 469.333s210.128 469.333 469.333 469.333c259.206 0 469.333-210.128 469.333-469.333v0c0.003-0.635 0.005-1.387 0.005-2.138 0-258.027-209.173-467.2-467.2-467.2-0.752 0-1.504 0.002-2.255 0.005l0.117-0zM512 896c-212.077 0-384-171.923-384-384s171.923-384 384-384c212.077 0 384 171.923 384 384v0c-1.203 211.592-172.408 382.797-383.885 383.999l-0.115 0.001z"
        />
        <path
          d="M512 426.667c-23.564 0-42.667 19.103-42.667 42.667v0 256c0 23.564 19.103 42.667 42.667 42.667s42.667-19.103 42.667-42.667v0-256c0-23.564-19.103-42.667-42.667-42.667v0z"
        />
        <path
          d="M554.667 298.667c0 23.564-19.103 42.667-42.667 42.667s-42.667-19.103-42.667-42.667c0-23.564 19.103-42.667 42.667-42.667s42.667 19.103 42.667 42.667z"
        />
      </svg>
    </span>
  </td>
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    2
  </td>
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    2
  </td>
  <td
    class="emotion-2"
    color="#3570f40D"
  />
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    <button
      class="emotion-6 denali-button"
      type="submit"
    >
      Update
    </button>
    
  </td>
</tr>
`;

exports[`TemplateRow should render with description 1`] = `
.emotion-1 {
  fill: #188fff;
  cursor: pointer;
  vertical-align: text-bottom;
}

.emotion-6 {
  border: none;
  border-radius: 2px;
  box-shadow: 0 0 0 1px transparent inset,0 0 0 0 #d5d5d5 inset;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  font-family: Helvetica,Arial,sans-serif;
  font-weight: 300;
  font-size: 12px;
  line-height: 1;
  margin: 5px;
  min-width: 70px;
  outline: 0;
  padding: 8px 18px;
  text-align: center;
  text-shadow: none;
  text-transform: none;
  -webkit-transition: all 0.2s ease-in;
  transition: all 0.2s ease-in;
  vertical-align: baseline;
  white-space: nowrap;
  background: transparent;
  box-shadow: 0 0 0 1px #3697f2 inset;
  color: #3697f2;
}

.emotion-6:first-of-type {
  margin-left: 0;
}

.emotion-6:disabled {
  background: rgba(48,48,48,0.1);
  color: rgba(48,48,48,0.2);
  cursor: not-allowed;
}

.emotion-6:disabled {
  box-shadow: 0 0 0 1px rgba(48,48,48,0.1) inset;
}

.emotion-6:hover:not(:disabled) {
  background: #d7e2fd;
}

.emotion-0 {
  background-color: #3570f40D;
  text-align: left;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
}

.emotion-2 {
  background-color: #3570f40D;
  text-align: center;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
}

<tr
  data-testid="template-row"
>
  <td
    class="emotion-0"
    color="#3570f40D"
  />
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    <span>
      <svg
        class="emotion-1"
        data-testid="icon"
        height="1.25em"
        viewBox="0 0 1024 1024"
        width="1.25em"
      >
        <title>
          information-circle
        </title>
        <path
          d="M512 42.667c-259.206 0-469.333 210.128-469.333 469.333s210.128 469.333 469.333 469.333c259.206 0 469.333-210.128 469.333-469.333v0c0.003-0.635 0.005-1.387 0.005-2.138 0-258.027-209.173-467.2-467.2-467.2-0.752 0-1.504 0.002-2.255 0.005l0.117-0zM512 896c-212.077 0-384-171.923-384-384s171.923-384 384-384c212.077 0 384 171.923 384 384v0c-1.203 211.592-172.408 382.797-383.885 383.999l-0.115 0.001z"
        />
        <path
          d="M512 426.667c-23.564 0-42.667 19.103-42.667 42.667v0 256c0 23.564 19.103 42.667 42.667 42.667s42.667-19.103 42.667-42.667v0-256c0-23.564-19.103-42.667-42.667-42.667v0z"
        />
        <path
          d="M554.667 298.667c0 23.564-19.103 42.667-42.667 42.667s-42.667-19.103-42.667-42.667c0-23.564 19.103-42.667 42.667-42.667s42.667 19.103 42.667 42.667z"
        />
      </svg>
    </span>
  </td>
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    2
  </td>
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    2
  </td>
  <td
    class="emotion-2"
    color="#3570f40D"
  />
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    <button
      class="emotion-6 denali-button"
      type="submit"
    >
      Update
    </button>
    
  </td>
</tr>
`;

exports[`TemplateRow should render without pop up 1`] = `
.emotion-1 {
  fill: #188fff;
  cursor: pointer;
  vertical-align: text-bottom;
}

.emotion-6 {
  border: none;
  border-radius: 2px;
  box-shadow: 0 0 0 1px transparent inset,0 0 0 0 #d5d5d5 inset;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  font-family: Helvetica,Arial,sans-serif;
  font-weight: 300;
  font-size: 12px;
  line-height: 1;
  margin: 5px;
  min-width: 70px;
  outline: 0;
  padding: 8px 18px;
  text-align: center;
  text-shadow: none;
  text-transform: none;
  -webkit-transition: all 0.2s ease-in;
  transition: all 0.2s ease-in;
  vertical-align: baseline;
  white-space: nowrap;
  background: transparent;
  box-shadow: 0 0 0 1px #3697f2 inset;
  color: #3697f2;
}

.emotion-6:first-of-type {
  margin-left: 0;
}

.emotion-6:disabled {
  background: rgba(48,48,48,0.1);
  color: rgba(48,48,48,0.2);
  cursor: not-allowed;
}

.emotion-6:disabled {
  box-shadow: 0 0 0 1px rgba(48,48,48,0.1) inset;
}

.emotion-6:hover:not(:disabled) {
  background: #d7e2fd;
}

.emotion-0 {
  background-color: #3570f40D;
  text-align: left;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
}

.emotion-2 {
  background-color: #3570f40D;
  text-align: center;
  padding: 5px 0 5px 15px;
  vertical-align: middle;
  word-break: break-all;
}

<tr
  data-testid="template-row"
>
  <td
    class="emotion-0"
    color="#3570f40D"
  />
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    <span>
      <svg
        class="emotion-1"
        data-testid="icon"
        height="1.25em"
        viewBox="0 0 1024 1024"
        width="1.25em"
      >
        <title>
          information-circle
        </title>
        <path
          d="M512 42.667c-259.206 0-469.333 210.128-469.333 469.333s210.128 469.333 469.333 469.333c259.206 0 469.333-210.128 469.333-469.333v0c0.003-0.635 0.005-1.387 0.005-2.138 0-258.027-209.173-467.2-467.2-467.2-0.752 0-1.504 0.002-2.255 0.005l0.117-0zM512 896c-212.077 0-384-171.923-384-384s171.923-384 384-384c212.077 0 384 171.923 384 384v0c-1.203 211.592-172.408 382.797-383.885 383.999l-0.115 0.001z"
        />
        <path
          d="M512 426.667c-23.564 0-42.667 19.103-42.667 42.667v0 256c0 23.564 19.103 42.667 42.667 42.667s42.667-19.103 42.667-42.667v0-256c0-23.564-19.103-42.667-42.667-42.667v0z"
        />
        <path
          d="M554.667 298.667c0 23.564-19.103 42.667-42.667 42.667s-42.667-19.103-42.667-42.667c0-23.564 19.103-42.667 42.667-42.667s42.667 19.103 42.667 42.667z"
        />
      </svg>
    </span>
  </td>
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    2
  </td>
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    2
  </td>
  <td
    class="emotion-2"
    color="#3570f40D"
  />
  <td
    class="emotion-2"
    color="#3570f40D"
  >
    <button
      class="emotion-6 denali-button"
      type="submit"
    >
      Update
    </button>
    
  </td>
</tr>
`;
